File: 프로그래머스/unrated/154538. 숫자 변환하기/숫자 변환하기.py
from collections import deque

def solution(x, y, n):
    dp = [0]*(y+1)
    q = deque([x])
    
    if x == y:
        return 0
    
    while q:
        v = q.popleft()
        for i in range(3):
            if i == 0:
                nv = v*2
            elif i == 1:
                nv = v*3
            else:
                nv = v+n
            
            if nv > y:
                continue
            if dp[nv]:
                continue
                
            if nv == y:
                return dp[v]+1
        
            q.append(nv)
            dp[nv] = dp[v]+1
    
    return -1

File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.java
import java.util.Deque;
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int solution(int x, int y, int n) {
        boolean[] visited = new boolean[y + 1];
        int count = 0;

        Deque<Integer> q = new LinkedList<>();
        q.push(x);
        visited[x] = true;

        if (x == y) {
            return 0;
        }

        while (!q.isEmpty()) {
            int size = q.size();

            for (int i = 0; i < size; i++) {
                int cur = q.poll();

                if (cur == y) {
                    return count;
                }

                addNext(cur + n, q, visited, y);
                addNext(cur * 2, q, visited, y);
                addNext(cur * 3, q, visited, y);
            }

            count += 1;
        }

        return -1;
    }

    public void addNext(int next, Queue<Integer> q, boolean[] visited, int limit) {
        if (next <= limit && !visited[next]) {
            q.offer(next);
            visited[next] = true;
        }
    }
}